### ISSUE TITLE
JSFX plugin format

### ISSUE THREAD

**Issue by jpcima** [falkTX/Carla#1487]
This is a source-based plugin format, which is made by Cockos and part of Reaper.
(Note that JSFX does not stand for JavaScript.)

There are dozens of these plugins shipped in Reaper (see `InstallData/Effects` folder).
Most are rather short source code files, tagged with a free software license on them.

The implementation of JSFX is partially free software.
[EEL2] is the underlying programming language, which is published as part of WDL.
There exists a free software implementation in [jsusfx].

Some notes
- the language would be useful for prototyping all sorts of dsp
- the files are arranged in a multi-level folder hierarchy. one might need to compile them in order to scan them
- the source is expected to compile at runtime to native assembly ; the build system is a little convoluted, which has architecture-dependent code and involves assembly in the nasm syntax.
- the jsusfx implementation is last updated 3 years ago, and the EEL2 is a bit behind; in particular, the later one has support of the arm 64bit processor.

[EEL2]: https://github.com/justinfrankel/WDL/tree/master/WDL/eel2
[jsusfx]: https://github.com/asb2m10/jsusfx/

**Comment by falkTX** [comment]
lets keep the discussion here then.
my main doubt against all this is a sorta "vendor lockin" if we place this in carla. It kinda makes sense for carla to load them since it is really a plugin format in the end, plus the drag&drop would make this handy... but it would be something limited to carla instead of being usable everywhere.

the dynamic nature of these files likely makes it very tricky to have as general plugins though, that part I do not know how to solve.

If carla really ends up supporting jsfx, then I would see it as a motivation to do a carla mini-host, a 1-to-1 wrapper for the things carla is able to load.

how fast is the scanning for these files?

**Comment by jpcima** [comment]
There is the current status at this branch.
Currently it builds the support libraries.
https://github.com/jpcima/Carla/tree/jsfx

**Comment by jpcima** [comment]
Note: there appears to be a mistake in file `jsusfx.cpp`
The following should be modified from `ifdef` to `ifndef`, to let scripts print their messages.
`#ifdef EEL_STRING_STDOUT_WRITE`
In addition, it should preferably use the carla-specific IO functions.

**Comment by jpcima** [comment]
[ ] Other: values don't parse correctly when `LC_NUMERIC != C`
Test case is Reaper `guitar/phaser`, which fails to modulate anything, because of multiplier `0.5` parsed as `0`.

**Comment by falkTX** [comment]
There is a class in carla utils for the locate issues, CarlaScopeLocale or something like that. either put it around the entire file loading or specific sections of it.

**Comment by jpcima** [comment]
Yes, this source code is C though. I'll fix it.
Culprit is here
https://github.com/falkTX/Carla/blob/555caea8a0e3aa28c1c1d9df081403b3efeadbe9/source/modules/eel2/source/WDL/eel2/nseel-compiler.c#L5869

**Comment by jpcima** [comment]
Re: carla-specific IO functions
The Jsfx output connects to a function which has the `fwrite` signature (it writes in chunks).
Hence it's not possible to connect with carla_stdout, as is?

**Comment by falkTX** [comment]
if it expects no newline, then no.
all the carla_std* functions append a newline at the end

**Comment by jpcima** [comment]
it's okay to do a light refactor of carla's io functions, to get access to that FILE handle for direct writing?

also `carla_stdout` and its friends have FILE* as a static-local into a static-inline function, it could be that these static handles end up multiple. (if multiple source units duplicate the function)

**Comment by jpcima** [comment]
- [ ] the files should be compiled as part of `init`, rather than `reload`
a recompilation may modify the number of i/o ports, which is not acceptable.

**Comment by jpcima** [comment]
Regarding author and category metadata

Many files contain comments similar to these ones.
```
//tags: guitar modulation filter gain
//author: Cockos
```

These look similar to the `desc:` tag except for being commented.
Most of the stock jsfx have these, yet it's not a standardized syntax afaik.
(it might be for a future extension of the language?)
For time being, there is absolutely a possibility of parsing these pseudo-tags.

**Comment by falkTX** [comment]
Makes sense to try to carefully parse them I think, they provide useful info.

btw, how do we differentiate between FX and synths?

**Comment by jpcima** [comment]
> btw, how do we differentiate between FX and synths?

- by category pseudo tags ("synthesis")
- absence of input pins
- possibly some light code analysis, for presence of midi calls

**Comment by jpcima** [comment]
- [ ] jsusfx does not support convolution plugins 

It does not support sliders written in that form, and the compilation/readHeader fails:
`slider1:/amp_models:none:Model`

The idea is that this slider (menu rather) controls an index of a file into a directory of wav IR files.
These wav files reside in the `Data/amp_models` folder, whereas the effect is located inside `Effects/guitar`.

The specification has something to say about this, more precisely: it scans wav/txt/ogg/raw files
It does not mention that files should be listed in lexicographical order, but in Reaper they are.

There is library support to implement, but also the pesky path search.
(since we have to go one above from `Effects` and into `Data`, such that `Data` is a hardcoded element of the path)

**Comment by jpcima** [comment]
- [ ] hidden sliders

The sliders whose desc begins with character '-' should be hidden parameters.
Specification says
> You can also hide sliders by prefixing their names with "-". Such parameters will not be visible in the plug-in UI but still be active, automatable, etc.

**Comment by jpcima** [comment]
- [ ] output-only sliders

There is this unspecified syntax, which creates a slider with empty range (min=max=0), and its purpose is to serve as output-only parameter. (but we don't normally know that unless examining the code)
`slider10:0,Reduction (dB)`

- [ ] slider automation from DSP

The observation related to the item above is that DSP can assign to a slider variable and have the slider move as a result.
Carla should detect the slider change, and automate the parameter.

**Comment by GavinRay97** [comment]
Just wanted to come and drop a huge note of appreciation and support for this!

Many JSFX are as good or better than paid plugins, and there's such a huge community collection of them.

Would be incredible to have a cross-platform implementation of JSFX host. 

Thank you for working on this =)

**Comment by jpcima** [comment]
> Would be incredible to have a cross-platform implementation of JSFX host.

There exists one as part the host support library. If you can build this, you get a vst plugin which can host jsfx.
https://github.com/jpcima/ysfx

I'll write a document later on how to build this.

**Comment by wolftune** [comment]
I can't get JSFX that have a UI to work. I used to succeed at getting ReaJS VST to load in Carla via Wine and show the UI. Now, that works for a moment and then it turns itself off.

It would be way way better if JSFX would just work directly *and* show the UI for each effect. I guess that's just not supported still in Carla…
